invenio_rdm_records/resources/serializers/schemas.py

```diff
@@ -6,12 +6,25 @@
 # it under the terms of the MIT License; see LICENSE file for more details.
 
 """Base parsing functions for the various serializers."""
-from marshmallow import Schema, fields, missing
+from marshmallow import Schema, missing, post_dump
+
+from invenio_rdm_records.contrib.journal.custom_fields import DumpProcessorMixin
 
 
 class BaseSchema(Schema):
     """BaseSchema for serializers in JSON."""
 
+    def __init__(self, processors=None, **kwargs):
+        """Initialize a new instance of the BaseSchema class.
+
+        :param processors: A list of `DumpProcessorMixin` instances used to modify the serialized output, defaults to None.
+        :type processors: list, optional
+        :raises AssertionError: If any item in the `processors` list is not an instance of `DumpProcessorMixin`.
+        """
+        super().__init__()
+        self.processors = processors or []
+        assert all(isinstance(p, DumpProcessorMixin) for p in self.processors)
```

Comment:
this is more of a moralistic quesiton: where is the threshold on using asserts instead of typing? my point being, if we do not use typing we assume the contract/programmatic API is respected/implemented. Then we do not have assertions. Otherwise, we should be consistent and assert all types of constuctor attributes (i.e. use typing)

In some very specific cases it makes sense to do assertions, what would be the reasoning behind this case?

Reply:
Nice question!

Just my 2 cents, feel free to disagree: I use `assert` when I expect an internal object to be as expected. If that is not the case, it means something is terribly broken in the system, internally.
For example, when I fetch a published record from the DB, I can assert that it does not have a `review`, see [here](https://github.com/inveniosoftware/invenio-rdm-records/blob/61d72f0e8f6c9e1c864548b61313a791e490b406/invenio_rdm_records/requests/community_inclusion.py#L42).

For value/type checking, I would prefer clear error msgs (you can do `assert <condition>, msg`), with an `if` with a `raise TypeError`, or even better, a specific excp. **Why?** Because a specific raised exception can then bubble up and can be caught by a `try/except`, and the code could recover in certain cases or behave/raise differently.

The other very minor difference is that assertions [might be disabled in Python if the code is optimized](https://discuss.python.org/t/stop-ignoring-asserts-when-running-in-optimized-mode/13132/41?page=3).

Honestly, in this specific case, I think it is fine to use `assert`: if the type is wrong, you cannot do much, it is an unrecoverable Internal Server Error.

Reply:
I'm consistently a pain regarding consistency haha

While I agree the it is nice to have a check to see if they implement the correct classes, the `processors` list comes form config (i.e. is not like a system check). Thus we are doing "parameter type checking", if we want that we should use Python's typing... Otherwise, implement asserts everywhere, for example:

- We should check custom fields implement BaseCF
- We should check PIDProviders implement PIDProvider
- a looong etc of classes or list of them that come from config (invenio.cfg) and can be changed by the user

I'm not against leaving the `assert`. I'm having a hard time understanding why in this case is needed but not in all the other one where a class or list of them come from config, so I would not know how to proceed if I had to code something similar.

>  The other very minor difference is that assertions might be disabled in Python if the code is optimized.

Nice, didnt know! Thanks for sharing.

Reply:
Just commenting to say that I heard typing and got interested.
I understand that this is/was something that was avoided so far, because the system was evolving too fast to be worth the effort... And if they get outdated regularly, they become unreliable and actually more harmful than beneficial.
Anyway, given that most of InvenioRDM's code is getting more and more stable (or at least that's what it feels like to me), would it make sense to reevaluate typing annotations?